packages/grid_client/src/clients/tf-grid/contracts.ts

```diff
@@ -301,13 +326,216 @@ class TFContracts extends Contracts {
+   * Function to convert USD to TFT
+   * @param {Decimal} USD the amount in USD.
+   * @returns  {Decimal} The amount in TFT.
+   */
+  private async convertToTFT(USD: Decimal) {
+    try {
+      const tftPrice = (await this.client.tftPrice.get()) ?? 0;
+      const tft = new currency(tftPrice).convertUSDtoTFT({ amount: USD.toNumber() });
+      return new Decimal(tft);
+    } catch (error) {
+      throw new GridClientError(`Failed to convert to mTFT due: ${error}`);
+    }
+  }
+  /**
+   * list all contracts, this is not restricted with the items counts
+   * this basically check if the count is larger than the page size, it make another request with the item count as the size pram.
+   * @param {GridProxyClient} proxy will be used to list the contracts
+   * @param {Partial<ContractsQuery>} queries
+   * @returns
+   */
+  async listAllContracts(proxy: GridProxyClient, queries: Partial<ContractsQuery>) {
+    const contracts = await proxy.contracts.list({ ...queries, retCount: true });
+    if (contracts.data.length < contracts.count!) {
+      return (
+        await proxy.contracts.list({
+          ...queries,
+          size: contracts.count!,
+        })
+      ).data;
+    } else return contracts.data;
+  }
+
+  /**
+   * Calculate total IPV4 and total overdraft on all node contracts on rented node.
+   * @description In node contracts on rented nodes we need to have both the overdraft to be added to rent contract overdraft.
+   * please note that the unbilled NU amount is added to the total overdraft exactly as the in the other contract types.
+   * the IPV4 cost to add to the estimated cost of the rent contract.
+   */
+  private async getContractsCostOnRentedNode(nodeId: number, proxy: GridProxyClient) {
+    const contracts = await this.listAllContracts(proxy, {
+      nodeId,
+      state: [ContractState.GracePeriod],
+      numberOfPublicIps: 1,
+    });
+
+    const ipPrice = (await this.client.pricingPolicies.get({ id: 1 })).ipu.value / 10 ** 7;
+    const BillingInformationPromises = contracts.reduce((acc: Promise<BillingInformation>[], contract) => {
+      acc.push(this.client.contracts.getContractBillingInformationByID(contract.contract_id));
+      return acc;
+    }, []);
+    const overDraftPromises = contracts.reduce((acc: Promise<ContractPaymentState>[], contract) => {
+      acc.push(this.client.contracts.getContractPaymentState(contract.contract_id));
+      return acc;
+    }, []);
+
+    const overDraftResult = await Promise.all(overDraftPromises);
+    const billingInfoResult = await Promise.all(BillingInformationPromises);
+
+    const totalOverDraft = overDraftResult.reduce(
+      (acc: Decimal, paymentState) => acc.add(paymentState.additionalOverdraft).add(paymentState.standardOverdraft),
+      new Decimal(0),
+    );
+    const totalNUCost = billingInfoResult.reduce((acc: number, billingInfo) => acc + billingInfo.amountUnbilled, 0);
+    const totalIPCost = (ipPrice * (contracts.data.length || 0)) / 1e7;
```

Comment:
By multiplying by `contracts.data.length`, are we assuming that a node contract can only have 1 IP address?

Reply:
the current behavior, yes, each contract should have 1 IP

Reply:
now I get the ip counts not the contracts